### PR TITLE
Add a few useful utility classes and functions

### PR DESCRIPTION
**Pull request by zhuzhile08**
Title pretty much says it all, there are some minor conflicts in the CMakeLists.txt in src, but rest is I think fine.
Recap, I've added:
 - an array wrapper
 - a scope based timer/benchmarker
 - a "dynamic" array (basically just a normal array with a vector interface)
 - a function pointer wrapper
 - a not thread safe logger
 - a resource/object pool implementation with a std-like interface
 - a smartpointer implementation
 - some random utility functions

### CODE REVIEW

File: include/utility/array.h [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+	typedef const value_type& const_reference;
+	typedef value_type array_type[Size];
+	typedef Array<value_type, Size> wrapper_type;
+
+	/**
+	 * @brief fill the array with the same value
+	 * 
+	 * @param value value to fill with
+	 */
+	void fill(const_reference value) { for (size_t i; i < Size; i++) m_array[i] = value; }
```
Comment (karnkaul) [inline include/utility/array.h:36]: Why are some member functions not `constexpr`?

File: include/utility/array.h [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+#include <utility>
+#include <algorithm>
+
+#define NODISCARD [[nodiscard]]
+#define DEPRECATED [[deprecated]]
+
+namespace utility {
+
+// basic C-array wrapper implementation
+template <class Ty, size_t Size> struct Array {
```
Comment (karnkaul) [inline include/utility/array.h:22]: Why is this needed when `std::array` exists and is in the PCH already?

File: include/utility/array.h [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+ * 
+ * @copyright Copyright (c) 2022
+ *************************/
+#pragma once
+
+#include <utility>
+#include <algorithm>
+
+#define NODISCARD [[nodiscard]]
+#define DEPRECATED [[deprecated]]
```
Comment (karnkaul) [inline include/utility/array.h:17]: These defines may already exist

Comment (karnkaul) [reply include/utility/array.h:17]: Why are they being used at all?

File: include/utility/dynarray.h [hunk 4]
```diff
@@ -0,0 +1,364 @@
...
+#pragma once
+
+#include <type_traits>
+#include <span>
+
+#include <utility/array.h>
+
+#define NODISCARD [[nodiscard]]
+#define DEPRECATED [[deprecated]]
+
```
Comment (karnkaul) [inline include/utility/dynarray.h:21]: Boom; include both array.h and dynarray.h and we get define clashes. Also, C++ headers should be `.hpp`.

File: include/utility/logger.h [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+	 * @param ...message messages
+	*/
+	template <typename ... Args> void debug(Args... message) {
+#ifndef NDEBUG
+		// set the color and font of the message
+		ANSI(Font::NON, Color::GRY);
+		// print the message
+		std::cout << "[DEBUG]: ";
+		(std::cout << ... << std::forward<Args>(message)) << end_l();
+#ifdef LYRA_LOG_FILE
```
Comment (karnkaul) [inline include/utility/logger.h:129]: What us `LYRA_LOG_FILE`?

File: include/utility/logger.h [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+	 * @tparam ...Args types to print
+	 * @param ...message messages
+	*/
+	template <typename ... Args> void debug(Args... message) {
+#ifndef NDEBUG
+		// set the color and font of the message
+		ANSI(Font::NON, Color::GRY);
+		// print the message
+		std::cout << "[DEBUG]: ";
+		(std::cout << ... << std::forward<Args>(message)) << end_l();
```
Comment (karnkaul) [inline include/utility/logger.h:128]: This isn't thread safe

File: include/utility/logger.h [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+#ifdef LYRA_LOG_FILE
+		m_logFile << "[DEBUG]: ";
+		(m_logFile << ... << std::forward<Args>(message)) << end_l();
+#endif
+		// reset color
+		set_color_default();
```
Comment (karnkaul) [inline include/utility/logger.h:134]: Also not thread safe

File: include/utility/array.h (hunk 2 above)
Comment (zhuzhile08) [reply include/utility/array.h:22]: Sorry, bad habit of mine, I like to use my own containers with a simpler and more readable implementation, which also means shorter headers, if you think I should still add this, I'd gladly redo some stuff here.

Well, not too sure about thy array is still in the PCHs, if again, if I should still include this, then I'd also create a PR to replace some PCHs

I guess that the same also goes for the smart pointer and the function pointer, but I have once benchmarked my smart pointer implementation to be just a tiny bit fasters than STD's and the function pointer should also perform better overall, again, because of the much simpler implementation. This one I haven't benchmarked though.

File: include/utility/array.h (hunk 3 above)
Comment (zhuzhile08) [reply include/utility/array.h:17]: Normally, I'd have a central header with some macros and aliases like these, so I guess I could do that here too. I use these because they just look better and the standard uses macros like these too.

File: include/utility/array.h (hunk 1 above)
Comment (zhuzhile08) [reply include/utility/array.h:36]: Didn't know void could be `constexpr`-ed, I'll add this in the rest of the API too.

File: include/utility/logger.h (hunk 6 above)
Comment (zhuzhile08) [reply include/utility/logger.h:128]: Yes, I am aware of this major flaw, but I think this should do just fine


For now at least...


Jokes aside, I'll probably make a thread safe version of this the next few days, so yeah.

File: include/utility/logger.h (hunk 5 above)
Comment (zhuzhile08) [reply include/utility/logger.h:129]: Crap, guess I forgot to take out some old macros, I'll just remove these all depending on if we want file logging or not.

File: include/utility/dynarray.h (hunk 4 above)
Comment (zhuzhile08) [reply include/utility/dynarray.h:21]: Alright, as mentioned above, I'll make that central header and change it all to `.hpp`, then I think this should be good to go.

**Comment by aleokdev** [comment]
No reinventing the c++ stdlib for any minor project challenge (impossible)
